### PR TITLE
General cleanup by making use of java 16 features

### PR DESCRIPTION
**Pull request by modmuss50**
Untested for now, just wanted to play with new toys

### CODE REVIEW

**Comment by modmuss50** [comment]
Uhh looks like gson doesnt play nicely with records, kinda expected...

**Comment by Toffikk** [comment]
It was not updated in more than a year

**Comment by modmuss50** [comment]
Yeah, need to find a good replacement. For now I think I can prob just whip up a custom TypeAdapterFactory.

Moshi seems to the go-to replacement but doesnt seem to support records right now from a quick google.

**Comment by modmuss50** [comment]
jackson-databind seems to do it quite nicely, I hope its not too bloated.

File: src/main/java/net/fabricmc/loom/api/decompilers/DecompilationMetadata.java [hunk 1]
```diff
@@ -27,14 +27,6 @@
...
-	public final Path javaDocs;
-	public final Collection<Path> libraries;
-
-	public DecompilationMetadata(int numberOfThreads, Path javaDocs, Collection<Path> libraries) {
-		this.numberOfThreads = numberOfThreads;
-		this.javaDocs = javaDocs;
-		this.libraries = libraries;
-	}
+public record DecompilationMetadata(int numberOfThreads, Path javaDocs,
+									Collection<Path> libraries) {
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/api/decompilers/DecompilationMetadata.java]: The indentation is really flying away here... I guess it looks better when tabs are shown as 4 spaces and not 8 (GH's default)?

File: src/main/java/net/fabricmc/loom/decompilers/fernflower/AbstractFernFlowerDecompiler.java [hunk 2]
```diff
@@ -65,12 +65,12 @@ public void decompile(Path compiledJar, Path sourcesDestination, Path linemapDes
...
 
-		Map<String, Object> options = new HashMap<String, Object>() {{
+		Map<String, Object> options = new HashMap<>() {{
 				put(IFernflowerPreferences.DECOMPILE_GENERIC_SIGNATURES, "1");
 				put(IFernflowerPreferences.BYTECODE_SOURCE_MAPPING, "1");
 				put(IFernflowerPreferences.REMOVE_SYNTHETIC, "1");
 				put(IFernflowerPreferences.LOG_LEVEL, "trace");
-				put(IFernflowerPreferences.THREADS, metaData.numberOfThreads);
+				put(IFernflowerPreferences.THREADS, metaData.numberOfThreads());
 			}};
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/decompilers/fernflower/AbstractFernFlowerDecompiler.java:74]: Hmm, the code format on the last line here looks a bit odd too

File: src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftVersionMeta.java [hunk 3]
```diff
@@ -29,108 +29,35 @@
...
-	public String getType() {
-		return type;
-	}
-
-	public final class AssetIndex extends Downloadable {
-		private String id;
-		private long totalSize;
-
-		public String getFabricId(String version) {
+public final record MinecraftVersionMeta(
```
Comment (Juuxel) [inline src/main/java/net/fabricmc/loom/configuration/providers/minecraft/MinecraftVersionMeta.java]:
```suggestion
public record MinecraftVersionMeta(
```
They're always final.

**Review by Juuxel — APPROVED** [review]

File: src/main/java/net/fabricmc/loom/api/decompilers/DecompilationMetadata.java (hunk 1 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/api/decompilers/DecompilationMetadata.java]: Ill just make this 1 line, blame idea for this.

File: src/main/java/net/fabricmc/loom/decompilers/fernflower/AbstractFernFlowerDecompiler.java (hunk 2 above)
Comment (modmuss50) [reply src/main/java/net/fabricmc/loom/decompilers/fernflower/AbstractFernFlowerDecompiler.java:74]: Yeah, this is checkstyle's doing. Might leave it for now

**Pull request merged**